protocols/kad/src/behaviour.rs

```diff
@@ -623,6 +731,93 @@ where
                 self.add_provider.push((key, provider_peer.node_id));
                 return;
             }
+            KademliaHandlerEvent::GetValue { key, request_id } => {
+                let (result, closer_peers) = match self.records.get(&key) {
+                    Some(record) => {
+                        (Some(record.into_owned()), Vec::new())
+                    },
+                    None => {
+                        let closer_peers = self.find_closest(&kbucket::Key::from(key), &source);
+                        (None, closer_peers)
+                    }
+                };
+
+                self.queued_events.push(NetworkBehaviourAction::SendEvent {
+                    peer_id: source,
+                    event: KademliaHandlerIn::GetValueRes {
+                        result,
+                        closer_peers,
+                        request_id,
+                    },
+                });
+            }
+            KademliaHandlerEvent::GetValueRes {
+                result,
+                closer_peers,
+                user_data,
+            } => {
+                let mut finished_query = None;
+
+                if let Some(query) = self.active_queries.get_mut(&user_data) {
+                    if let QueryInfoInner::GetValue {
+                        key: _,
+                        results,
+                        num_results,
+                    } = &mut query.target_mut().inner {
+                        if let Some(result) = result {
+                            results.push(result);
+                            if results.len() == *num_results {
+                                finished_query = Some(user_data);
+                            }
+                        }
+                    }
+                }
+
+                if let Some(finished_query) = finished_query {
+                    let (query_info, _) = self
+                        .active_queries
+                        .remove(&finished_query)
+                        .expect("finished_query was gathered when peeking into active_queries; QED.")
+                        .into_target_and_closest_peers();
+
+                    match query_info.inner {
+                        QueryInfoInner::GetValue { key: _, results, .. } => {
+                            let result = GetValueResult::Found { results };
+                            let event = KademliaOut::GetValueResult(result);
+
+                            self.queued_events.push(NetworkBehaviourAction::GenerateEvent(event));
+                        }
+                        _ => panic!("unexpected query_info.inner variant for a get_value result; QED.")
```

Comment:
This is still not a good proof. It does indicate what is wrong, but the point is to prove that things cannot be wrong.
I didn't intent that you try to fix it now, and fixing it probably requires code changes.

```suggestion
                        // TODO: write a better proof
                        _ => panic!("unexpected query_info.inner variant for a get_value result; QED.")
```